### ISSUE TITLE
Allow prepared statements with PDO

### ISSUE THREAD

**Issue by psychoticmeow** [symphonycms/symphony-2#2425]
The `Database` class has switched to using PDO internally, but it doesn't expose PDOs prepared statements. Prepared statements allow safer and more powerful querying, and their usage should be encouraged.

**Comment by nitriques** [comment]
+1

**Comment by brendo** [comment]
Agree, the first step of `PDO` has been done, which was just getting it in. Now that's done, we can try and really turn the screws to start realising the benefits of `PDO`. How do you want to go from here?

**Comment by psychoticmeow** [comment]
I'm going to implement `Database::prepare`. It'll have to translate table name prefixes but I think that's it really.

**Comment by psychoticmeow** [comment]
What is the minimum PHP version for Symphony 3?

**Comment by brendo** [comment]
Good question. The realistic release for Symphony 3 is probably August/September, so it'd make sense for us to mirror PHP.net's support as well. In the past we supported PHP 5.2 (and even now, PHP 5.3) for far too long simply because of poor adoption rates amongst hosts. I think the landscape has changed in recent years and there's a push for hosts to 'be better'. 

Referring to [PHP.net supported versions](http://php.net/supported-versions.php) page, I think a sensible target would be PHP 5.5+ as both PHP 5.3 (already dead) and PHP 5.4 will have reached end of life.

How's that sound? If hosts don't support it, well, I think people should start to apply the pressure :+1:

**Comment by psychoticmeow** [comment]
It really depends on when you expect to release it, and how that sits with what PHP is doing right now.

1. There are no plans for PHP 5.7, PHP 5.6 will be the last PHP 5 release.
2. PHP 7 RCs are expected to start landing next month.
3. According to the only [source of statistics](http://w3techs.com/technologies/details/pl-php/5/all) I could find more than 40% of people are still running PHP 5.3.

Also the two largest server distributions are both going to be out of date:

1. Ubuntu 14.04 LTS has PHP 5.5 and the next LTS isn't until 16.04.
2. RHEL 7.1 (Just released) has PHP 5.4 and likely won't update for at least a year

I think that realistically users of both distros are going to have to install alternative PHP packages in the near future to get the most recent version, so I suggest we just worry about what we need for development.

**Comment by jonmifsud** [comment]
I'm mostly on VPN's nowadays so can make the move up. I guess it will also
put pressure for people to keep updating their client's sites not keep
running a legacy symphony version because they can have 5.3 running on all
systems.

**Comment by jensscherbl** [comment]
> How's that sound? If hosts don't support it, well, I think people should start to apply the pressure

I already switched my [hosting provider][1] for new projects since my [old provider][2] only started "upgrading" to PHP 5.4 recently, while most other providers offer PHP 5.6

[1]: http://www.strato.de
[2]: https://www.hosteurope.de

**Comment by munki-boy** [comment]
Isn't it a PITA if you had to move a ton of small sites from one shared hosting to another? One I use called Heart Internet are good shared hosts and seem to upgrade PHP but what if they stop :panicface

There's a lot of small businesses round here and it's nicer for me to use Symphony than it is needed for their site functionality. 

I mean might you make a situation where there's a bunch of people stuck on earlier symphony?

**Comment by psychoticmeow** [comment]
@munki-boy I'm sure a bunch of sites will be stuck on an earlier version of Symphony, it already happens because available extensions change over time.

That said, it's important to update PHP once the security patches for the version you're using stop. Older Symphony sites should be forwards compatible with the newest version.

And it's also important to improve our code with the improvements that PHP 5.5 and 5.6 offer.

**Comment by munki-boy** [comment]
@psychoticmeow I had a look and I can choose up-to PHP 5.5 on my hosting at the moment so it looks like they're keeping up.  My clients own hosting thinks 5.4 is bleeding edge so that'll have to stay on the old versions. I use few extensions that aren't core so that's not too bad for me.

**Comment by nitriques** [comment]
I would say php 5.4+, sadly.

**Comment by psychoticmeow** [comment]
So... we're going to target PHP 5.4, a release that is about to run out of support for a project that's due to be released when exactly?

We should at least be targeting 5.5, it brings [a lot of improvements](https://wiki.php.net/rfc#php_55), but [generators](https://wiki.php.net/rfc/generators) alone would make it worth our while.

**Comment by designermonkey** [comment]
IMO, we as a team can no longer afford to 'play it safe' and target an end of life PHP version. I run a lot of servers now, and the ones that are PHP 5.4 and lower keep me awake at night as they **are no longer receiving security updates**. 

As a software package, we should be pushing forwards and targeting the minimum of 5.6. By the time Sym 3.0.0 is released, we should probably be targeting higher than that.

I'm sorry to say, and I don't mean to aggravate anyone here, but if you are concerned about not being able to run Symphony on a shared host because they only have 5.4, then your concerns should be with the hosting company not being willing to secure your application for you by providing a patchable environment. Move hosts.

**Comment by designermonkey** [comment]
Remember too guys, Symphony 2 is a great piece of software. We should support it for security for `x` number of years, but Symphony 3 needs to move on, and adopt new paradigms. There will be big changes in the core; There already are, and it will need a more robust future proofed version of PHP behind it.

Please don't hold the scope of what can be done back by worrisome pandering to bad shared hosts. They are killing the PHP market, and we should try to be part of the modern PHP movement by pushing past these dinosaurs and adopting some of the amazing advances in PHP.

**Comment by michael-e** [comment]
As far as I have seen, even Debian (which is surely not the fastest distro in adopting new versions) has moved to 5.6 with Debian 8 (Jessie), so I agree.

**Comment by froschdesign** [comment]
> As far as I have seen, even Debian has moved to 5.6 with Debian 8 (Jessie)

Correct: https://packages.debian.org/en/jessie/php5

**Comment by jensscherbl** [comment]
> we should be pushing forwards and targeting the minimum of 5.6.

:+1: 

> if you are concerned about not being able to run Symphony on a shared host because they only have 5.4, then your concerns should be with the hosting company not being willing to secure your application for you by providing a patchable environment. Move hosts.

:+1:

**Comment by jonmifsud** [comment]
I'm on the same lines with John - I don't think Symphony 3 has to stay
stuck with php 5.4, anyway it's not like there's a release tomorrow. + I
think anyone would owe it to their clients especially if you're including
'security updates' to be on a decent version of php.

**Comment by brendo** [comment]
Yeah, as I mentioned originally, I think PHP 5.5+ minimum. For those who can't upgrade the environment, don't upgrade to Symphony 3. We got burnt hard by supporting PHP 5.2 for as long as we did, it'd be nice to avoid that again.

**Comment by munki-boy** [comment]
I'm the serial whiner about shared hosting and I agree now, having had the situation explained the other day. It's some clients that have their own shared hosting that'd be a problem but you're right, if they want more features they can upgrade. 

I must say though, "just move hosts" is a bit easy sounding; but there's around 100 domains here on a shared host with several email accounts each. It'd be a job of work moving it right?

**Comment by designermonkey** [comment]
it would need planning, definitely. I had to move an entire datacenter not to long ago. It took two weeks with all the DNS changes, prepping the new servers etc. If they're all on the same server, not too big a job, just plan DNS switches carefully.

**Comment by munki-boy** [comment]
Cheers John, I'd started wondering if I was missing some secret shortcut.

**Comment by munki-boy** [comment]
It's a clients own shared hosting account, not like me moving some of my own clients around where I could do it case by case.

**Comment by nitriques** [comment]
You all have valid points and did convince me to move forward. I guess it will be time to buy new servers then ;)

What was bugging me the most was that some old version of Symphony (Remember 2.1, 2.2) could not run on newer php versions.

If we decide to follow php's newest versions, we should make sure that the maintenance release makes it possible to run them on the new versions as well.

**Comment by michael-e** [comment]
> If we decide to follow php's newest versions, we should make sure that the maintenance release makes it possible to run them on the new versions as well.

+1. BTW, does Sym 2.6.x run flawlessly on PHP 5.6?

**Comment by nitriques** [comment]
> does Sym 2.6.x run flawlessly on PHP 5.6?

From what's I've tested yes. But was never able to run 2.2.x with php5.3 and do not want that again.
